File: src/main/ssh-auth.ts
import { existsSync, readFileSync } from 'fs'
import { homedir, userInfo } from 'os'
import { join } from 'path'

export interface ResolvedSsh {
  host: string
  port: number
  username: string
  identityFiles: string[]
}

interface HostBlock {
  patterns: string[]
  hostName?: string
  user?: string
  port?: number
  identityFiles: string[]
}

export function sshDir(): string {
  return join(homedir(), '.ssh')
}

export function agentPath(): string | undefined {
  if (process.platform === 'win32') {
    const pipe = '\\\\.\\pipe\\openssh-ssh-agent'
    return existsSync(pipe) ? pipe : undefined
  }
  const sock = process.env.SSH_AUTH_SOCK
  return sock && existsSync(sock) ? sock : undefined
}

export function defaultIdentityFiles(): string[] {
  const dir = sshDir()
  return ['id_ed25519', 'id_ecdsa', 'id_rsa', 'id_dsa']
    .map((name) => join(dir, name))
    .filter((p) => existsSync(p))
}

export function resolveSshTarget(host: string, username?: string, port?: number): ResolvedSsh {
  const blocks = loadSshConfig()
  const matched = blocks.filter((b) => b.patterns.some((p) => matchHost(p, host)))
  const hostName = first(matched, (b) => b.hostName) || host
  const user = username || first(matched, (b) => b.user) || userInfo().username || 'root'
  const resolvedPort = port || first(matched, (b) => b.port) || 22
  const identityFiles = unique([
    ...matched.flatMap((b) => b.identityFiles),
    ...defaultIdentityFiles()
  ].map(expandPath).filter((p) => existsSync(p)))

  return { host: hostName, port: resolvedPort, username: user, identityFiles }
}

function first<T, R>(items: T[], pick: (item: T) => R | undefined): R | undefined {
  for (const item of items) {
    const value = pick(item)
    if (value !== undefined && value !== '') return value
  }
  return undefined
}

function unique(items: string[]): string[] {
  const seen = new Set<string>()
  const out: string[] = []
  for (const item of items) {
    const key = item.toLowerCase()
    if (seen.has(key)) continue
    seen.add(key)
    out.push(item)
  }
  return out
}

function expandPath(p: string): string {
  if (p.startsWith('~/') || p.startsWith('~\\')) return join(homedir(), p.slice(2))
  if (p === '~') return homedir()
  return p.replace(/%d/g, homedir()).replace(/%u/g, userInfo().username)
}

function matchHost(pattern: string, host: string): boolean {
  const negated = pattern.startsWith('!')
  const p = negated ? pattern.slice(1) : pattern
  const re = new RegExp('^' + p.replace(/[.+^${}()|[\]\\]/g, '\\$&').replace(/\*/g, '.*').replace(/\?/g, '.') + '$', 'i')
  const ok = re.test(host)
  return negated ? !ok : ok
}

function loadSshConfig(): HostBlock[] {
  const file = join(sshDir(), 'config')
  if (!existsSync(file)) return []
  let text = ''
  try {
    text = readFileSync(file, 'utf8')
  } catch {
    return []
  }

  const blocks: HostBlock[] = []
  let current: HostBlock | null = null
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.replace(/#.*$/, '').trim()
    if (!line) continue
    const sep = line.search(/\s|=/)
    if (sep < 0) continue
    const key = line.slice(0, sep).toLowerCase()
    const value = line.slice(sep + 1).trim().replace(/^"|"$/g, '')
    if (key === 'host') {
      current = { patterns: value.split(/\s+/).filter(Boolean), identityFiles: [] }
      blocks.push(current)
      continue
    }
    if (!current) continue
    if (key === 'hostname') current.hostName = value
    else if (key === 'user') current.user = value
    else if (key === 'port') current.port = Number(value) || 22
    else if (key === 'identityfile') current.identityFiles.push(value)
  }
  return blocks
}

export function isAuthFailure(err: unknown): boolean {
  const m = (err instanceof Error ? err.message : String(err)).toLowerCase()
  return (
    m.includes('auth') ||
    m.includes('permission denied') ||
    m.includes('unable to authenticate') ||
    m.includes('all configured authentication methods failed') ||
    m.includes('encrypted private') ||
    m.includes('passphrase') ||
    m.includes('no matching') ||
    m.includes('login')
  )
}

export function isHostUnreachable(err: unknown): boolean {
  const m = (err instanceof Error ? err.message : String(err)).toLowerCase()
  const code = (err as NodeJS.ErrnoException).code || ''
  return (
    code === 'ECONNREFUSED' ||
    code === 'ENOTFOUND' ||
    code === 'EHOSTUNREACH' ||
    code === 'ETIMEDOUT' ||
    m.includes('timed out') ||
    m.includes('timeout') ||
    m.includes('econnrefused') ||
    m.includes('enotfound') ||
    m.includes('ehostunreach')
  )
}

export class SshConnectError extends Error {
  constructor(
    message: string,
    readonly code: 'AUTH' | 'HOST' | 'UNKNOWN',
    readonly keysTried: string[] = []
  ) {
    super(message)
    this.name = 'SshConnectError'
  }
}

export function isAgentFailure(err: unknown): boolean {
  const m = (err instanceof Error ? err.message : String(err)).toLowerCase()
  const code = (err as NodeJS.ErrnoException).code || ''
  return (
    code === 'ENOENT' ||
    m.includes('agent') ||
    m.includes('pageant') ||
    m.includes('named pipe')
  )
}

File: src/CompareShow/Core/SshAuth.cs
using System.Text.RegularExpressions;

namespace CompareShow.Core;

public sealed class ResolvedSsh
{
    public string Host { get; set; } = "";
    public int Port { get; set; } = 22;
    public string Username { get; set; } = "";
    public List<string> IdentityFiles { get; set; } = [];
}

public sealed class SshConnectException : Exception
{
    public string Code { get; }
    public IReadOnlyList<string> KeysTried { get; }

    public SshConnectException(string message, string code, IReadOnlyList<string>? keysTried = null)
        : base(message)
    {
        Code = code;
        KeysTried = keysTried ?? [];
    }
}

public static class SshAuth
{
    public static string SshDir() => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".ssh");

    public static IEnumerable<string> DefaultIdentityFiles()
    {
        var dir = SshDir();
        foreach (var name in new[] { "id_ed25519", "id_ecdsa", "id_rsa", "id_dsa" })
        {
            var p = Path.Combine(dir, name);
            if (File.Exists(p)) yield return p;
        }
    }

    public static ResolvedSsh Resolve(string host, string? username = null, int? port = null)
    {
        var blocks = LoadSshConfig();
        var matched = blocks.Where(b => b.Patterns.Any(p => MatchHost(p, host))).ToList();
        var hostName = matched.Select(b => b.HostName).FirstOrDefault(v => !string.IsNullOrEmpty(v)) ?? host;
        var user = username
            ?? matched.Select(b => b.User).FirstOrDefault(v => !string.IsNullOrEmpty(v))
            ?? Environment.UserName
            ?? "root";
        var resolvedPort = port ?? matched.Select(b => b.Port).FirstOrDefault(v => v is > 0) ?? 22;

        var files = new List<string>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var f in matched.SelectMany(b => b.IdentityFiles).Concat(DefaultIdentityFiles()))
        {
            var expanded = ExpandPath(f);
            if (!File.Exists(expanded) || !seen.Add(expanded)) continue;
            files.Add(expanded);
        }

        return new ResolvedSsh { Host = hostName, Port = resolvedPort, Username = user, IdentityFiles = files };
    }

    public static bool IsAuthFailure(Exception err)
    {
        var m = err.Message.ToLowerInvariant();
        return m.Contains("auth")
            || m.Contains("permission denied")
            || m.Contains("unable to authenticate")
            || m.Contains("all configured authentication methods failed")
            || m.Contains("encrypted private")
            || m.Contains("passphrase")
            || m.Contains("no matching")
            || m.Contains("login")
            || m.Contains("denied");
    }

    public static bool IsHostUnreachable(Exception err)
    {
        var m = err.Message.ToLowerInvariant();
        return err is System.Net.Sockets.SocketException
            || m.Contains("timed out")
            || m.Contains("timeout")
            || m.Contains("refused")
            || m.Contains("unreachable")
            || m.Contains("no such host")
            || m.Contains("could not resolve");
    }

    sealed class HostBlock
    {
        public List<string> Patterns { get; set; } = [];
        public string? HostName { get; set; }
        public string? User { get; set; }
        public int? Port { get; set; }
        public List<string> IdentityFiles { get; set; } = [];
    }

    static List<HostBlock> LoadSshConfig()
    {
        var file = Path.Combine(SshDir(), "config");
        if (!File.Exists(file)) return [];
        string text;
        try { text = File.ReadAllText(file); }
        catch { return []; }

        var blocks = new List<HostBlock>();
        HostBlock? current = null;
        foreach (var raw in text.Split('\n'))
        {
            var line = Regex.Replace(raw, @"#.*$", "").Trim();
            if (line.Length == 0) continue;
            var sep = line.IndexOfAny([' ', '=', '\t']);
            if (sep < 0) continue;
            var key = line[..sep].ToLowerInvariant();
            var value = line[(sep + 1)..].Trim().Trim('"');
            if (key == "host")
            {
                current = new HostBlock { Patterns = value.Split([' ', '\t'], StringSplitOptions.RemoveEmptyEntries).ToList() };
                blocks.Add(current);
                continue;
            }
            if (current is null) continue;
            if (key == "hostname") current.HostName = value;
            else if (key == "user") current.User = value;
            else if (key == "port") current.Port = int.TryParse(value, out var p) ? p : 22;
            else if (key == "identityfile") current.IdentityFiles.Add(value);
        }
        return blocks;
    }

    static bool MatchHost(string pattern, string host)
    {
        var negated = pattern.StartsWith('!');
        var p = negated ? pattern[1..] : pattern;
        var re = "^" + Regex.Escape(p).Replace("\\*", ".*").Replace("\\?", ".") + "$";
        var ok = Regex.IsMatch(host, re, RegexOptions.IgnoreCase);
        return negated ? !ok : ok;
    }

    static string ExpandPath(string p)
    {
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        if (p == "~") return home;
        if (p.StartsWith("~/") || p.StartsWith("~\\")) return Path.Combine(home, p[2..]);
        return Environment.ExpandEnvironmentVariables(p.Replace("%d", home).Replace("%u", Environment.UserName));
    }
}
